src/main/java/com/google/devtools/build/lib/remote/http/HttpCacheClient.java

```diff
@@ -588,85 +583,97 @@ public void close() {
           }
         };
     UploadCommand upload = new UploadCommand(uri, casUpload, key, wrappedIn, length);
-    Channel ch = null;
-    boolean success = false;
     if (storedBlobs.putIfAbsent((casUpload ? CAS_PREFIX : AC_PREFIX) + key, true) == null) {
-      try {
-        ch = acquireUploadChannel();
-        ChannelFuture uploadFuture = ch.writeAndFlush(upload);
-        uploadFuture.sync();
-        success = true;
-      } catch (Exception e) {
-        // e can be of type HttpException, because Netty uses Unsafe.throwException to re-throw a
-        // checked exception that hasn't been declared in the method signature.
-        if (e instanceof HttpException) {
-          HttpResponse response = ((HttpException) e).response();
-          if (authTokenExpired(response)) {
-            refreshCredentials();
-            // The error is due to an auth token having expired. Let's try again.
-            if (!reset(in)) {
-              // The InputStream can't be reset and thus we can't retry as most likely
-              // bytes have already been read from the InputStream.
-              throw e;
-            }
-            putAfterCredentialRefresh(upload);
-            success = true;
-            return;
-          }
-        }
-        throw e;
-      } finally {
-        if (!success) {
-          storedBlobs.remove(key);
-        }
-        in.close();
-        if (ch != null) {
-          releaseUploadChannel(ch);
-        }
-      }
+      SettableFuture<Void> result = SettableFuture.create();
+      acquireUploadChannel()
+          .addListener(
+              (Future<Channel> chP) -> {
```

Comment:
nit: `channelPromise`?

Reply:
Changed. I didn't change the download side.